Classes/Expression.lua

```diff
@@ -0,0 +1,230 @@
+=
+math.abs, math.acos, math.asin, math.atan, math.atan2,
+math.ceil, math.cos, math.cosh, math.exp, math.floor, math.log,
+math.log, math.log10, math.max, math.min, math.round, math.sin,
+math.sinh, math.sqrt, math.tan, math.tanh, math.random, math.pi, math.exp(1)
+
+-- These functions are not build into Lua:
+local cbrt = function(x) return sqrt(x^(1/3)) end
+local randint = function(max) return random(0, max) end
+-- TODO: rint function
+
+%s
+
+return function(%s)
+	%s
+	return %s
+end]]
+
+
+
+
+
+-- The envoronment of the loader. 
+-- It can currently only use the functions from the math library.
+cExpression.m_LoaderEnv =
+{
+	math = math,
+}
+
+
+
+
+
+-- All the assignment operator
+-- Since Lua only supports the simple = assignments we need to give the others special handling
+cExpression.m_Assignments =
+{
+	"=",
+	"%+=",
+	"%-=",
+	"%*=",
+	"%%=",
+	"%^=",
+	"/=",
+}
+
+
+
+
+
+-- A list of all the comparison operators. This is used to see if an action is an assignment or a comparison.
+-- For example if "x=5;y<z" was given as input then the first action is an assignment, while the second action is a comparison.
+cExpression.m_Comparisons =
+{
+	"<",
+	">",
+	"<=",
+	">=",
+	"==",
+	"!=",
+	"~=",
+}
+
+
+
+
+
+function cExpression:new(a_Formula)
+	local Obj = {}
+	
+	a_Formula = a_Formula
+	:gsub("!=", "~=") -- Lua operator for not equal is ~=
+	:gsub("&&", " and ")
+	:gsub("||", " or ")
+	
+	setmetatable(Obj, cExpression)
+	self.__index = self
+	
+	-- The string of the formula
+	Obj.m_Formula = a_Formula
+	
+	-- All the variables that that the formula can use. For example x, y and z
+	Obj.m_Parameters = {}
+	
+	-- All the variables the formula will return after executing
+	Obj.m_ReturnValues = {}
+	
+	-- A table containing predefined variables. A new one can be added using the PredefineVariable function
+	Obj.m_PredefinedConstants = {}
+	
+	return Obj
+end
+
+
+
+
+
+-- Adds a new parameter to the formula. The formula can use this in the calculation.
+-- a_Name is a string that will be the name of the parameter
+function cExpression:AddParameter(a_Name)
+	table.insert(self.m_Parameters, a_Name)
+	return self
+end
+
+
+
+
+
+-- Makes the formula return a variable when executing
+-- a_Name is the name of the variable that will be returned.
+function cExpression:AddReturnValue(a_Name)
+	table.insert(self.m_ReturnValues, a_Name)
+	return self
+end
+
+
+
+
+
+-- Adds a new constant. The formula will be able to use this in it's calculation.
+-- a_VarName is a string that will be the name of the constant.
+-- a_Value can only be a string or a number, since the environment blocks all other functions and tables.
+function cExpression:PredefineConstant(a_VarName, a_Value)
+	table.insert(self.m_PredefinedConstants, {name = a_VarName, value = a_Value})
+	return self
+end
+
+
+
+
+
+-- Creates a safe function from the formula string, the bound parameters and the predefined variables.
+function cExpression:Compile()
```

Comment:
The returned function takes the previously-bound parameters (AddParameter()), does the calculations using any predefined constants (PredefineConstant()) and returns the named values (AddReturnValue())